File: app/src/main/java/org/biotstoiq/gophercle/URL.java
package org.biotstoiq.gophercle;

public class URL {

    static String url;
    static String protocol;
    static String host;
    static int port;
    static String path;
    static String query;
    static char itemType;

    static int errorCode;

    static boolean urlOkay;

    URL(String input) {
        url = input;
        urlOkay = extractURLParts();
    }

    boolean extractURLParts() {
        int index;

        if(url.contains("://")) {
            index = url.indexOf(':');
            protocol = url.substring(0, index);
            if(!protocol.equals("gopher")) {
                errorCode = 1;
                return false;
            }
            url = url.substring(index+ + 3);
        } else {
            protocol = "gopher";
        }

        index = url.indexOf(":");
        if(index != -1) {
            host = url.substring(0, index);
            boolean extractedPort;
            if(url.contains("/")) {
                extractedPort = extractPort(url.substring(index + 1, index = url.indexOf("/")));
                url = url.substring(index);
            } else {
                extractedPort = extractPort(url.substring(index + 1));
                url = "";
            }
            if (!extractedPort) {
                errorCode = 3;
                return false;
            }
        } else {
            port = 70;
            index = url.indexOf("/");
            if(index != -1) {
              host = url.substring(0, index);
              url = url.substring(index);
            } else {
              host = url;
              url = "";
            }
        }

        if(url.length() > 1) {
          url = url.substring(1);
          index = url.indexOf('/');
          if(index == 1) {
            path = url.substring(index);
          } else {
            path = "/".concat(url);
          }
        } else {
          path = "/";
        }

        index = path.indexOf('\t');
        if (index != -1) {
            query = path.substring(index);
        } else {
            query = "";
        }
        return true;
    }

    void makeURLfromParts() {
        url = protocol.concat("://").concat(host).concat(":")
                .concat(String.valueOf(port).concat("/"))
                .concat(String.valueOf(itemType)).concat(path);
    }

    boolean extractPort(String string) {
        int port;
        try {
            port = Integer.parseInt(string);
        } catch(NumberFormatException | NullPointerException e) {
            e.printStackTrace();
            return false;
        }
        URL.port = port;
        return true;
    }

    int getErrorCode() { return errorCode; }
    void setErrorCode(int ec) { errorCode = ec; }

    String getUrl() {
        return url;
    }

    boolean isUrlOkay() { return urlOkay; }

    String getUrlHost() { return host; }
    void setUrlHost(String urlHst) { host = urlHst; }

    int getUrlPort() { return port; }
    void setUrlPort(int urlPrt) { port = urlPrt; }

    String getUrlPath() { return path; }
    void setUrlPath(String urlPth) { path = urlPth; }

    String getUrlQuery() { return query; }

    char getUrlItemType() { return itemType; }
    void setUrlItemType(char itmTyp) { itemType = itmTyp; }
}

File: app/src/main/java/org/biotstoiq/gophercle/URL.kt
package org.biotstoiq.gophercle

@Suppress("SpellCheckingInspection")
class URL internal constructor(url: String) {
    init {
        isUrlOkay = extractURLParts()
    }

    fun extractURLParts(): Boolean {
        var index: Int
        if (url.contains("://")) {
            index = url.indexOf(':')
            protocol = url.substring(0, index)
            if (protocol != "gopher") {
                errorCode = 1
                return false
            }
            url = url.substring(index + 3)
        } else {
            protocol = "gopher"
        }
        index = url.indexOf(":")
        if (index != -1) {
            urlHost = url.substring(0, index)
            val extractedPort: Boolean
            if (url.contains("/")) {
                extractedPort = extractPort(url.substring(index + 1, url.indexOf("/").also { index = it }))
                url = url.substring(index)
            } else {
                extractedPort = extractPort(url.substring(index + 1))
                url = ""
            }
            if (!extractedPort) {
                errorCode = 3
                return false
            }
        } else {
            urlPort = 70
            index = url.indexOf("/")
            if (index != -1) {
                urlHost = url.substring(0, index)
                url = url.substring(index)
            } else {
                urlHost = url
                url = ""
            }
        }
        if (url.length > 1) {
            url = url.substring(1)
            index = url.indexOf('/')
            urlPath = if (index == 1) {
                url.substring(index)
            } else {
                "/" + url
            }
        } else {
            urlPath = "/"
        }
        index = urlPath!!.indexOf('\t')
        urlQuery = if (index != -1) {
            urlPath!!.substring(index)
        } else {
            ""
        }
        return true
    }

    fun makeURLfromParts() {
        url = protocol + "://" + urlHost + ":" + urlPort.toString() + "/" + urlItemType.toString() + urlPath
    }

    fun extractPort(string: String): Boolean {
        val port: Int
        port = try {
            string.toInt()
        } catch (e: NumberFormatException) {
            e.printStackTrace()
            return false
        } catch (e: NullPointerException) {
            e.printStackTrace()
            return false
        }
        urlPort = port
        return true
    }

    var errorCode: Int = 0
    var url: String = ""
        private set

    companion object {
        var protocol: String? = null
        var urlHost: String? = null
            private set
        var urlPort: Int = 0
            private set
        var urlPath: String? = null
            private set
        var urlQuery: String? = null
            private set
        var urlItemType: Char = 0.toChar()
            private set
        var isUrlOkay: Boolean = false
            private set
    }
}
